larsim/EventGenerator/MultiPart/MultiPartRain_module.cc

```diff
@@ -0,0 +1,547 @@
+    if (param.kerange[0] > param.kerange[1]) this->abort("KE range has no phase space...");
+
+    if (_debug > 0) {
+      std::cout << "Generating particle (PDG";
+      for (auto const& pdg : param.pdg)
+        std::cout << " " << pdg;
+      std::cout << ")" << std::endl
+                << (param.use_mom ? "    KE range ....... " : "    Mom range ....... ")
+                << param.kerange[0] << " => " << param.kerange[1] << " MeV" << std::endl
+                << std::endl;
+    }
+
+    _param_v.push_back(param);
+  }
+}
+
+void MultiPartRain::beginRun(art::Run& run)
+{
+  // grab the geometry object to see what geometry we are using
+  art::ServiceHandle<geo::Geometry> geo;
+
+  std::unique_ptr<sumdata::RunData> runData(new sumdata::RunData(geo->DetectorName()));
+
+  run.put(std::move(runData), art::fullRun());
+
+  return;
+}
+
+std::vector<size_t> MultiPartRain::GenParticles() const
+{
+
+  std::vector<size_t> result;
+  std::vector<size_t> gen_count_v(_param_v.size(), 0);
+  std::vector<double> weight_v(_param_v.size(), 0);
+  for (size_t idx = 0; idx < _param_v.size(); ++idx)
+    weight_v[idx] = _param_v[idx].weight;
+
+  int num_part = (int)(fFlatRandom->fire(_multi_min, _multi_max + 1 - 1.e-10));
+
+  while (num_part) {
+
+    double total_weight = 0;
+    for (auto const& v : weight_v)
+      total_weight += v;
+
+    double rval = 0;
+    rval = fFlatRandom->fire(0, total_weight);
+
+    size_t idx = 0;
+    for (idx = 0; idx < weight_v.size(); ++idx) {
+      rval -= weight_v[idx];
+      if (rval <= 0.) break;
+    }
+
+    // register to the output
+    result.push_back(idx);
+
+    // if generation count exceeds max, set probability weight to be 0
+    gen_count_v[idx] += 1;
+    if (gen_count_v[idx] >= _param_v[idx].multi[1]) weight_v[idx] = 0.;
+
+    --num_part;
+  }
+  return result;
+}
+
+void MultiPartRain::GenPosition(double& x, double& y, double& z)
+{
+
+  x = fFlatRandom->fire(_xrange[0], _xrange[1]);
+  y = fFlatRandom->fire(_yrange[0], _yrange[1]);
+  z = fFlatRandom->fire(_zrange[0], _zrange[1]);
+
+  if (_debug > 0) {
+    std::cout << "Generating a rain particle at (" << x << "," << y << "," << z << ")" << std::endl;
+  }
+}
+
+std::array<double, 3U> MultiPartRain::extractDirection() const
+{
+  double px, py, pz;
+  if (_cosmic_distribution) {
+    double phi = fFlatRandom->fire(0, 2 * util::pi());
+    // Zenith Angle Theta is in [pi/2, pi]
+    double costheta = -1. * fCosmicAngleRandom->fire();
+    double sintheta = sqrt(1 - costheta * costheta);
+    pz = cos(phi) * sintheta;
+    px = sin(phi) * sintheta;
+    py = costheta;
+  }
+  else {
+    px = fNormalRandom->fire(0, 1);
+    py = fNormalRandom->fire(0, 1);
+    pz = fNormalRandom->fire(0, 1);
+    double p = std::hypot(px, py, pz);
+    px = px / p;
+    py = py / p;
+    pz = pz / p;
+  }
+
+  std::array<double, 3U> result = {px, py, pz};
+  return result;
+}
+
+void MultiPartRain::GenMomentum(const PartGenParam& param,
```

Comment:
See comment above about changing the return type.  Probably the function signature should look like:

```c++
TVector3 MultiPartRain::GenMomentum(PartGenParam const& param, double mass, TVector3 const& pos);
```